Feature/Send Token UI

bhumitaylor14 opened the pull request:
## Description
Add send native token UI 

## Motivation and Context
Develop Etherspot UI library for sending native token.

## How Has This Been Tested?
On local machine

## Screenshots (if appropriate):
<img width="1786" alt="Screenshot 2023-12-18 at 11 21 39 AM" src="https://github.com/etherspot/ui/assets/141008551/1acb7ba2-4801-404d-8f99-9acd4a6f0b94">
<img width="660" alt="Screenshot 2023-12-18 at 11 51 09 AM" src="https://github.com/etherspot/ui/assets/141008551/f28858ef-2432-4499-bdeb-1547f00b8630">

<img width="855" alt="Screenshot 2023-12-18 at 5 19 28 PM" src="https://github.com/etherspot/ui/assets/141008551/db350af6-89de-4026-aaee-d2f49206fe38">

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ tick] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

bhumitaylor14 committed:
WIP - add send native token UI

bhumitaylor14 committed:
validation of receiver address and trggerElement

bhumitaylor14 committed:
WIP | tailwind configuration

bhumitaylor14 committed:
WIP | integrating button for check tailwind css

bhumitaylor14 committed:
add send Native token UI with styling

bhumitaylor14 committed:
snapshot testing

bhumitaylor14 committed:
WIP | send logic

bhumitaylor14 committed:
Call send and estimate using Transaction kit

IAmKio commented on __tests__/SendNativeToken.test.tsx:
```diff
@@ -0,0 +1,58 @@
+
+  // Assert: call function of send
+  expect(inputElement.value).toBe('0.001');
+});
```
Remember new line at the end, do you have EditorConfig installed? This will add a new line automatically to all files

IAmKio commented on example/README.md:
```diff
@@ -0,0 +1,70 @@
+# Getting Started with Create React App
```
Please update this README with what the example app is and what someone can expect to find when they run it

IAmKio commented on example/package.json:
```diff
@@ -0,0 +1,46 @@
+  "dependencies": {
+    "@emotion/react": "^11.10.5",
+    "@emotion/styled": "^11.10.5",
+    "@etherspot/etherspot-ui": "file:./../",
```
Remember to add instructions in the README in regards to running rollup first otherwise this won't work. Also shouldn't this be:

```suggestion
    "@etherspot/etherspot-ui": "file:./../dist",
```
or something?

IAmKio commented on example/public/index.html:
```diff
@@ -0,0 +1,43 @@
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
+    <title>Send Native Token Example</title>
```
Remember this is the example app for all widgets, not just Send Native Token

IAmKio commented on example/public/logo192.png:
We'll update this and the app manifest. See attached ZIP
[IconKitchen-Output (7).zip](https://github.com/etherspot/ui/files/13726466/IconKitchen-Output.7.zip)

IAmKio commented on example/public/manifest.json:
Updated manifest.json in the ZIP file attached to the comment above

IAmKio commented on example/package.json:
```diff
@@ -0,0 +1,46 @@
+    "@emotion/styled": "^11.10.5",
+    "@etherspot/etherspot-ui": "file:./../",
+    "@mui/lab": "^5.0.0-alpha.118",
+    "@mui/material": "^5.11.7",
```
Are these actually being used yet?

IAmKio commented on example/src/App.tsx:
```diff
@@ -0,0 +1,28 @@
+  </div>
+  );
+};
+export default App;
```
Newline, please remember to add .editorconfig file

IAmKio commented on example/src/index.tsx:
Are you using a formatter here? There are extra / missing linebreaks everywhere

IAmKio commented on postcss.config.js:
```diff
@@ -0,0 +1,6 @@
+    tailwindcss: {},
+    autoprefixer: {},
+  },
+}
```
New line

IAmKio commented on rollup.config.js:
```diff
@@ -66,12 +73,16 @@ export default [
       'etherspot',
       '@etherspot/prime-sdk',
       '@etherspot/transaction-kit',
+      
```
Is this new line meant to be here?

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+import React, {useRef} from 'react';
```
Missing the Etherspot MIT License:

https://github.com/etherspot/skandha/blob/master/LICENSE

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+import React, {useRef} from 'react';
+import   SendNativeTokenUI   from './sendNativeTokenUI';
```
There are two spaces around `SendNativeTokenUI`

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+import {
+  EtherspotTransactionKit
+} from "@etherspot/transaction-kit";
+import { ethers } from "ethers";
```
Please categorise these imports

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+  onlyEstimate?: boolean;
+}
+
+const SendNativeToken = ({
```
Can you please add JSDOC reference here for this function

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+  onlyEstimate = false,
+}: SendNativeTokenProps) => {
+  const randomWallet = ethers.Wallet.createRandom();
+  const providerWallet = new ethers.Wallet(randomWallet.privateKey);
```
Is this supposed to be here?

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+  console.log(providerWallet);
+  return (
+  <EtherspotTransactionKit
+    provider={providerWallet} /* The random wallet we created above */
```
We need to accept the provider from the props that we make available to the developer, not the random wallet

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -0,0 +1,54 @@
+        style={style}
+        triggerElement={triggerElementRef}
+      />
+     </EtherspotTransactionKit>
```
Just out of curiosity, why have we chosen to wrap `SendNativeTokenUI` in `SendNativeToken`? Could these two not be merged into one?

IAmKio commented on src/components/SendNativeToken/index.tsx:
Please annotate this whole file and explain what it is doing at each code block

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+import React, { useState, KeyboardEvent, useEffect } from 'react';
```
Add Etherspot MIT license, linked above

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+  useEtherspotTransactions,
+} from "@etherspot/transaction-kit";
+import { ethers } from "ethers";
+import { isValidEthereumAddress } from '../../utils/validation';
```
Please categorise these imports

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+  triggerElement,
+  disableSendOnEnter = false,
+  onlyEstimate = false,
+}: SendNativeTokenProps) => {
```
Add JSDOC reference here for this function

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+  const [isValidAddress, setIsValidAddress] = useState(true);
+  const [error, setError] = useState('');
+  const { estimate, send } = useEtherspotTransactions();
+  const randomWallet = ethers.Wallet.createRandom();
+  const providerWallet = new ethers.Wallet(randomWallet.privateKey);
+  console.log(providerWallet);
```
Random wallet is here again - please remove this and accept a provider that is passed via props from the developer

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+    };
+  }, [triggerElement]);
+
+  useEffect(() => {
+    // validate receiver address
+    if(isValidEthereumAddress(receiverAddress)){  
+      setIsValidAddress(true);
+    }else{
+      setIsValidAddress(false);
+    }
+  }, [receiverAddress]);
```
Please check formatting

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+    setAmount(event.target.value);
+  };
+
+  // logic of estimate and send based on props
+  /*const estimateAndSend = async () => {
+    if (isValidAddress) {
+      // Estimate logic
+      const gasEstimation = await transactionKitProps.provider.estimateGas({
+        to: transactionKitProps.to,
+        value: transactionKitProps.amount,
+      });
+  
+      if (!onlyEstimate) {
+        // Send logic
+        try {
+          const transactionHash = await transactionKitProps.provider.sendTransaction({
+            to: transactionKitProps.to,
+            value: transactionKitProps.amount,
+            gasLimit: gasEstimation.toNumber() + 10000, // Adding some buffer to the estimated gas limit
+          });
+  
+          console.log('Transaction sent! Transaction Hash:', transactionHash);
+          // Optionally, you can handle the success of the transaction here
+        } catch (error) {
+          console.error('Error sending transaction:', error);
+          // Optionally, you can handle the failure of the transaction here
+        }
+      }
+  
+      if (debug) {
+        console.log('send');
+      }
+    } else {
+      setError('Receiver address is not a valid blockchain address');
+    }
+  }; */
```
Please remove dead code

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+      }
+  
+      if (debug) {
+        console.log('send');
```
Please made debug messages more descriptive...

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+      <EtherspotBatches>
+        <EtherspotBatch>
+          <EtherspotTransaction to={transactionKitProps.to} value={transactionKitProps.amount}>
+            <div className="mt-2.5">
```
Should this be here?

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,163 @@
+  
+};
+
+export default SendNativeTokenUI;
```
New line

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
Remember to annotate the whole file. Also please rename `sendNativeTokenUI.tsx` to `SendNativeTokenUI.tsx`

IAmKio replied:
This was not completed: 

> Also please rename sendNativeTokenUI.tsx to SendNativeTokenUI.tsx

IAmKio commented on src/utils/validation.ts:
```diff
@@ -0,0 +1,12 @@
+  
+    try {
+      return ethers.utils.isAddress(address);
+    } catch (e) {
+      //
+    }
```
There's an empty catch block here?

IAmKio commented on src/utils/validation.ts:
```diff
@@ -0,0 +1,12 @@
+      //
+    }
+    return false;
+  };
```
New line

IAmKio commented on tailwind.config.js:
```diff
@@ -0,0 +1,9 @@
+{import('tailwindcss').Config}
+module.exports = {
+  content: ["./src/**/*.{js,ts,jsx,tsx}", "./node_modules/YOUR_LIBRARY/**/*.{js,ts,jsx,tsx}"], 
```
Just wondering what happens with `YOUR_LIBRARY` here?

IAmKio reviewed (changes requested):
Please check the comments. In particular:

- Coding style should be consistent everywhere
- All code should be annotated
- Use EditorConfig to keep line endings etc consistent
- Ensure each component file has the Etherspot MIT license at the top before any code
- Ensure there is a LICENSE file within the repository with the same Etherspot MIT license
- The provider should be provided by the developer, not a random wallet

Please remember that this code will be public and used by other developers, remember to make this a developer first experience and they will look at this code also. Bad code makes for bad first impressions.

bhumitaylor14 committed:
Cosmetic changes

IAmKio commented on example/src/App.tsx:
```diff
@@ -1,11 +1,19 @@
+import { useRef } from 'react';
+import { ethers } from "ethers"; 
+
+//Components
```
Please add a space between // and Components

IAmKio commented on example/src/App.tsx:
```diff
@@ -18,11 +26,15 @@ const App = () => {
\ No newline at end of file
+
+export default App;
+
```
There's an extra line break here?

IAmKio commented on example/src/index.tsx:
```diff
@@ -1,12 +1,13 @@
 import React from 'react';
 import ReactDOM from 'react-dom/client';
 
+// components
```
Please keep the capitalisation consistent

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -1,10 +1,37 @@
+/* MIT License
+ * 
+ * Copyright (c) [Year] [Author]
```
You have left the Year and Author blank

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -15,8 +42,32 @@ interface SendNativeTokenProps {
+ * @property {boolean} [disableSendOnEnter=false] - Flag indicating whether to disable sending on Enter key press.
+ * @property {boolean} [onlyEstimate=false] - Flag indicating whether to only estimate the transaction.
+ * @property {Object | undefined} [provider] - Ethereum provider (random wallet provider).
+ */
```
Should this not be moved to before the interface?

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -15,8 +42,32 @@ interface SendNativeTokenProps {
+ * @description SendNativeToken component for sending native tokens using the EtherspotTransactionKit.
+ * @param {SendNativeTokenProps} props - The props for the component
+ * @returns {React.ReactElement} The rendered component.
+ */
```
This should sit directly above the function definition

IAmKio commented on src/components/SendNativeToken/index.tsx:
```diff
@@ -27,16 +78,20 @@ const SendNativeToken = ({
-  >
+    // EtherspotTransactionKit component, wrapping the SendNativeTokenUI
+    <EtherspotTransactionKit
+      provider={provider} /* The random wallet we created above */
```
You left the old comment here

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -1,25 +1,58 @@
+/* MIT License
+ * 
+ * Copyright (c) [Year] [Author]
```
You forgot to add the Year and Author

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -1,25 +1,58 @@
-}
+/**
+ * SendNativeTokenUI component for sending native tokens using the EtherspotTransactionKit.
+ *
```
Missing a `@name` and `@description` tags

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -1,25 +1,58 @@
+ * @property {React.ReactNode | React.RefObject<HTMLElement>} [triggerElement] - Trigger element for UI interactions.
+ * @property {boolean} [disableSendOnEnter=false] - Flag indicating whether to disable sending on Enter key press.
+ * @property {boolean} [onlyEstimate=false] - Flag indicating whether to only estimate the transaction.
+ */
```
You've added a JSDOC block here but i can't see the code?

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -56,108 +91,67 @@ const SendNativeTokenUI = ({
   const estimateAndSend = async () => {
     if (isValidAddress) {
       // Estimate logic
       const gasEstimation = await estimate();
```
What happens when this returns an error state?

IAmKio replied:
When debug mode is on, the gas estimation object should be logged

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -1,25 +1,58 @@
-  onlyEstimate?: boolean;
-}
+/**
+ * SendNativeTokenUI component for sending native tokens using the EtherspotTransactionKit.
```
This needs to be more detailed. Explain this as if you're explaining to someone who knows nothing about sending a native crypto token to another address. This is who this library is aimed at.

IAmKio reviewed (changes requested):
Several issues are still left unresolved. Specifically:

- Did you see my question about why we're using two files for one component?
- Should we not be pointing the example app to the the dist folder?
- License issues
- Code description issues

bhumitaylor14 committed:
Add code description and Other Cosmetic changes

bhumitaylor14 committed:
Removed unnecessary content from tailwind config

bhumitaylor14 commented:
@IAmKio ,
Did you see my question about why we're using two files for one component? 
- I have put comments below your comments and also annotate file with same description.
Should we not be pointing the example app to the the dist folder?
- @etherspot/etherspot-ui": "file:./../,  it is pointing dist folder only.
License issues 
- Updated year and author
Code description issues
- Updated description at necessary place.

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -48,8 +48,11 @@ import { isValidEthereumAddress } from '../../utils/validation';
 /**
- * SendNativeTokenUI component for sending native tokens using the EtherspotTransactionKit.
- *
+ * @name SendNativeToken
+ * @description SendNativeTokenUI component provides a UI for sending native crypto tokens
```
Naming mismatch here

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -48,8 +48,11 @@ import { isValidEthereumAddress } from '../../utils/validation';
- *
+ * @name SendNativeToken
+ * @description SendNativeTokenUI component provides a UI for sending native crypto tokens
+ * to another Ethereum address using the EtherspotTransactionKit.This user-friendly interface 
```
Please add a space after "." and "This"

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -48,8 +48,11 @@ import { isValidEthereumAddress } from '../../utils/validation';
+ * @description SendNativeTokenUI component provides a UI for sending native crypto tokens
+ * to another Ethereum address using the EtherspotTransactionKit.This user-friendly interface 
+ * is designed to simplify the complex underlying transaction mechanics, making it accessible 
+ * to users who may not be familiar with the intricacies of blockchain transactions.
```
Add a new line here for readability within the JSDOC block

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -114,20 +117,26 @@ const SendNativeTokenUI = ({
+
+        if (debug) {
+          console.log('Gas estimation:', gasEstimation);
+          console.log('Transaction sent successfully');
```
This console.log always assumes that we are going to send...

IAmKio commented on src/components/SendNativeToken/sendNativeTokenUI.tsx:
```diff
@@ -114,20 +117,26 @@ const SendNativeTokenUI = ({
-      setError('Receiver address is not a valid blockchain address');
+    } catch (error) {
+      // Handle errors from estimate or send methods
+      setError('Something wrong is happening to estimate and send transaction. Try again.');
```
We can't give errors like this, where we're intentionally vague about what the error _could_ be

- It's either an estimation OR send error
- Explain properly what the user needs to do to "try again"
- Add a debug flag and a console.error which outputs the error
- The use of a wide try / catch block is making this difficult to differentiate which error this is
- Also add a link to the eEherspot documentation at the end of the console.error

IAmKio commented on src/utils/validation.ts:
```diff
@@ -6,7 +6,7 @@ export const isValidEthereumAddress = (address: string | undefined): boolean =>
       return ethers.utils.isAddress(address);
     } catch (e) {
-      //
+      console.log(e);
```
Should this not be a console.error with a description with the error object?

IAmKio reviewed (changes requested):
Please see the comments. Also please do not self-resolve my comments as this is my comment that i need to check.

bhumitaylor14 committed:
Debug and error log changes

bhumitaylor14 committed:
Delete src/components/SendNativeToken/sendNativeTokenUI.tsx

bhumitaylor14 committed:
Capitalized file name

IAmKio commented on src/components/SendNativeToken/SendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,178 @@
+}: SendNativeTokenUIProps): React.ReactElement => {
+
+  // State for the amount input
+  const [amount, setAmount] = useState('0.001');
```
Should this be set to 0?

IAmKio commented on src/components/SendNativeToken/SendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,178 @@
+  // State for the amount input
+  const [amount, setAmount] = useState('0.001');
+  // State for checking if the receiver address is valid
+  const [isValidAddress, setIsValidAddress] = useState(true);
```
This should be `false` by default, we can't assume that the default input address is correct

IAmKio commented on src/components/SendNativeToken/SendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,178 @@
+      if (debug) {
+        console.error('Error in estimateAndSend:', error);
+  
+        if (error instanceof SomeSpecificEstimationError) {
+          console.error('Specific error in gas estimation:', error);
+        } else if (error instanceof SomeSpecificSendError) {
+          console.error('Specific error in sending transaction:', error);
+        }
```
What is `SomeSpecificEstimationError` and `SomeSpecificSendError`?

IAmKio replied:
@bhumitaylor14 stlll unresolved

IAmKio commented on src/components/SendNativeToken/SendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,178 @@
+    <EtherspotBatches>
+      <EtherspotBatch>
+        <EtherspotTransaction to={receiverAddress} value={amount}>
+          <div className="mt-2.5">
```
No defined class names here, that is for the developer to decide

IAmKio commented on src/components/SendNativeToken/SendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,178 @@
+              className={unstyled ? '' : className}
+              style={style}
+            />
+            {error ? <p className='text-sm'>{error}</p> : ''}
```
No class names to define style, the developer should decide this. We'll need to devise a better system for this in the next iteration, but for now because you've added another HTML component here, we need to cater for a developer wanting to style this. 

- Please add another prop to the component called `errorMessageClass` and make it available to this HTML element to allow a developer to be able to style it.

IAmKio commented on src/components/SendNativeToken/SendNativeTokenUI.tsx:
```diff
@@ -0,0 +1,178 @@
+      setError(
+        `An error occurred during the transaction. Please try again.${
+          debug ? ' Check the console for more details.' : ''
+        } If the issue persists, refer to the Etherspot documentation: https://etherspot.dev/docs`
```
This part of the message "If the issue persists, refer to the Etherspot documentation: https://etherspot.dev/docs" should be moved to inside the debug === true clause, as this is a developer focused message not a user message

IAmKio commented on src/utils/validation.ts:
```diff
@@ -0,0 +1,13 @@
+    if (!address) return false;
+  
+    try {
+      return ethers.utils.isAddress(address);
```
Does `ethers.utils.isAddress` actually throw an error or does it just return `true` / `false`? https://docs.ethers.org/v6/api/address/#isAddress it looks like the return type is just a boolean...

IAmKio replied:
Still unresolved

bhumitaylor14 committed:
changes suggested by Kio

bhumitaylor14 committed:
remove catch block as ethers.utils has already handled

bhumitaylor14 committed:
added custom class declaration

bhumitaylor14 committed:
remove custom error handling and make it simple

bhumitaylor14 committed:
using error instance to log errors

bhumitaylor14 committed:
error log change
